Run NetFramework tests inside vstest.console.exe process

Faizan2304 opened the pull request:
Run test within vstest.console.exe process for project targeting  `NetFramework` if running from commandline.
This will be the default behavior. To opt out default behavior user has to explicitly pass '/InIsolation' argument.

Commit:
No isolation discovery for net46 with design mode off

Commit:
NoIsolation test run for net46

Faizan2304 committed:
Merge branch 'master' into noIsolation

Faizan2304 committed:
Make api asynchronous

Commit:
Merge branch 'master' into noIsolation

Commit:
1) Renamed argument processor to InProcess
2) Wrote unit tests.

Faizan2304 commented:
Please see the matrix below to see the benefit of using `InProcess` argument.

### Performance matrix
| Number of tests        | Normal Flow (in seconds) | InProcess (in seconds)| Percentage benefit |
|------------------------| ---------------------------- |-------------------------|----------------------|
| 10 tests in one project | 1.20 seconds | 0.42 seconds | 65%|
|100 tests in one project | 1.23 seconds | 0.45 seconds | 63.4 %|
|1000 tests in one project| 1.73 seconds | 0.76 seconds |56%|
|10000 tests in one project | 9.15 seconds | 4.76 seconds | 47.9%|
|10000 tests in 10 projects | 9.91 seconds | 6.42 seconds | 35.2% |
|1 lakh test in 10 projects | 75.60 seconds | 46.60 seconds | 38.3%|

mayankbansal018 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs:
```diff
@@ -0,0 +1,127 @@
+            {
+                EqtTrace.Error("InProcessProxyexecutionManager.StartTestRun: Failed to start test run: {0}", exception);
+
+                // Send a discovery complete to caller.
```
nit: change comment to test execution complete

Faizan2304 replied:
Fixed in next commit.

mayankbansal018 commented on src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs:
```diff
@@ -59,14 +63,19 @@ public IProxyDiscoveryManager GetDiscoveryManager(ITestRuntimeProvider testHostM
         {
             var parallelLevel = this.VerifyParallelSettingAndCalculateParallelLevel(discoveryCriteria.Sources.Count(), discoveryCriteria.RunSettings);
```
this can be moved below

Faizan2304 replied:
Fixed in next commit.

mayankbansal018 commented on src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs:
```diff
@@ -193,5 +207,30 @@ private int VerifyParallelSettingAndCalculateParallelLevel(int sourceCount, stri
+            }
+
+            var runConfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(runsettings);
+
```
Add a comment specifying when will we enable InProc run

Faizan2304 replied:
Fixed

smadala commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs:
```diff
@@ -0,0 +1,101 @@
+        /// <param name="eventHandler">EventHandler for handling discovery events from Engine</param>
+        public void DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler eventHandler)
+        {
+            var discoveryManager = this.testHostManagerFactory.GetDiscoveryManager();
```
Why calling `GetDiscoveryManager()` in every method, Can be a member variable?

chakradhar2 replied:
and we can call Initilize() instead of doing check below.

Faizan2304 replied:
Fixed

smadala commented on test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/InProcessProxyDiscoveryManagerTests.cs:
```diff
@@ -0,0 +1,138 @@
+        public void InitializeShouldCallDiscoveryManagerInitializeWithEmptyIEnumerable()
+        {
+            this.inProcessProxyDiscoveryManager.Initialize();
+            this.mockDiscoveryManager.Verify(o => o.Initialize(Enumerable.Empty<string>()), Times.Once);
```
Add assert failure message for all asserts.

chakradhar2 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscoveryManager.cs:
```diff
@@ -60,8 +60,11 @@ public void Initialize(IEnumerable<string> pathToAdditionalExtensions)
-            // Start using these additional extensions
-            TestPluginCache.Instance.DefaultExtensionPaths = pathToAdditionalExtensions;
+            if (pathToAdditionalExtensions != null && pathToAdditionalExtensions.Count() > 0)
+            {
```
Can we use pathToAdditionalExtensions.Any() instead of pathToAdditionalExtensions.Count() > 0?

Faizan2304 replied:
Yes we can use

chakradhar2 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs:
```diff
@@ -0,0 +1,101 @@
+                    EqtTrace.Error("InProcessProxyDiscoveryManager.DiscoverTests: Failed to discover tests: {0}", exception);
+
+                    // Send a discovery complete to caller.
+                    eventHandler.HandleLogMessage(TestMessageLevel.Error, exception.Message);
```
And we can just call Initialize() instead of  doing check here. It seems its not being used any where else.

Faizan2304 replied:
fixed

chakradhar2 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs:
```diff
@@ -0,0 +1,101 @@
+        /// </summary>
+        public void Abort()
+        {
+            Task.Run(() => this.testHostManagerFactory.GetDiscoveryManager().Abort());
```
Do we require any null check for  this.testHostManagerFactory?

Faizan2304 replied:
No, we don't require. If its null, process should catch exception

Faizan2304 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs:
```diff
@@ -0,0 +1,101 @@
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine.TesthostProtocol;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+    class InProcessProxyDiscoveryManager : IProxyDiscoveryManager
```
internal

Faizan2304 commented on src/Microsoft.TestPlatform.ObjectModel/RunSettings/RunConfiguration.cs:
```diff
@@ -376,6 +399,15 @@ public bool DisableAppDomainSet
         /// <summary>
+        /// Gets a value indicating whether InProcess is set.
+        /// </summary>
+        public bool InProcessSet
```
Removed

Faizan2304 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs:
```diff
@@ -0,0 +1,127 @@
+            }
+        }
+
+        public int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler)
```
fixed

Faizan2304 commented on src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs:
```diff
@@ -193,5 +207,30 @@ private int VerifyParallelSettingAndCalculateParallelLevel(int sourceCount, stri
+                !(runConfiguration.TargetFrameworkVersion.Name.IndexOf("netstandard", StringComparison.OrdinalIgnoreCase) >= 0
+                || runConfiguration.TargetFrameworkVersion.Name.IndexOf("netcoreapp", StringComparison.OrdinalIgnoreCase) >= 0))
+            {
+                return true ;
```
Fixed!

Commit:
1) Making in process default for net46
2) Enable InIsolation argument to disable default in process.
3) Address PR comment

Commit:
Merge branch 'master' into noIsolation

Commit:
Disable inprocess for inProcDataCollector.

codito commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs:
```diff
@@ -0,0 +1,96 @@
+
+                    // Send a discovery complete to caller.
+                    eventHandler.HandleLogMessage(TestMessageLevel.Error, exception.ToString());
+                    eventHandler.HandleDiscoveryComplete(-1, new List<TestCase>(), true);
```
`Enumerable.Empty<TestCase>`

Faizan2304 replied:
Fixed

codito commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs:
```diff
@@ -0,0 +1,126 @@
+                var executionContext = new TestExecutionContext(
+                            testRunCriteria.FrequencyOfRunStatsChangeEvent,
+                            testRunCriteria.RunStatsChangeEventTimeout,
+                            inIsolation: false,
```
How is this used in TPv1? Are we missing scenario?

Faizan2304 replied:
This info get passed to adapter through RunContext

codito commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs:
```diff
@@ -0,0 +1,126 @@
+
+                if (testRunCriteria.HasSpecificSources)
+                {
+                    // [TODO]: we need to revisit to second-last argument if we will enable datacollector. 
```
nit: remove todo

Faizan2304 replied:
Fixed

codito commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyexecutionManager.cs:
```diff
@@ -0,0 +1,126 @@
+                            hasTestRun: true,
+                            isDebug: (testRunCriteria.TestHostLauncher != null && testRunCriteria.TestHostLauncher.IsDebug),
+                            testCaseFilter: testRunCriteria.TestCaseFilter);
+
```
We should ask TestRuntimeProvider for extensions.

Faizan2304 replied:
Yes, we should.

Faizan2304 commented on src/Microsoft.TestPlatform.CrossPlatEngine/Client/InProcessProxyDiscoveryManager.cs:
```diff
@@ -0,0 +1,96 @@
+    internal class InProcessProxyDiscoveryManager : IProxyDiscoveryManager
+    {
+        private ITestHostManagerFactory testHostManagerFactory;
+        IDiscoveryManager discoveryManager;
```
Fixed in next commit.

Commit:
Addressed Arun comment

Commit:
Append comment

Commit:
Merge branch 'master' into noIsolation
